Fixes: register_full_backward_hook crash if first argument don't require a gradient (#57944)

sigmunjr opened the pull request:
Fixes #57944

Commit:
register_full_backward_hook crash if first argument don't require grad (#57944)

albanD reviewed (commented):
LGTM thanks for the quick fix!
Could you add one test in `test/test_nn.py` that makes sure we don't break this again in the future?

Commit:
Adding test to ensure no regression (#57944)

Adding a spesific to check that a hook does not crash when first input does not requires gradient.

albanD commented on test/test_nn.py:
```diff
@@ -622,6 +622,16 @@ def hook(mod, grad_input, grad_output):
+        try:
+            mod(inp.detach(), inp)
+        except Exception as ex:
+            self.fail("Unexpected exception: %s" % ex)
```
For code that should not raise exceptions, we usually don't protect it and add a comment saying so. But this works as well.

albanD reviewed (approved):
Thanks for the test!

Yes very specific is good for this :)
